doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -344,85 +343,96 @@ Rewriting with definitional equality
-   strategy_flag ::= {+ @red_flag }
-   | @delta_flag
-   red_flag ::= beta
-   | iota
-   | match
-   | fix
-   | cofix
-   | zeta
-   | delta {? @delta_flag }
-   reference_occs ::= @reference {? at @occs_nums }
-   pattern_occs ::= @one_term {? at @occs_nums }
-
-This set of tactics implements different specialized usages of the
-tactic :tacn:`change`.
-
-All conversion tactics (including :tacn:`change`) can be parameterized by the
-parts of the goal where the conversion can occur. This is done using
-*goal clauses* which consists in a list of hypotheses and, optionally,
-of a reference to the conclusion of the goal. For defined hypothesis
-it is possible to specify if the conversion should occur on the type
-part, the body part or both (default).
-
-Goal clauses are written after a conversion tactic (tactics :tacn:`set`,
-:tacn:`rewrite`, :tacn:`replace` and :tacn:`autorewrite` also use goal
-clauses) and are introduced by the keyword `in`. If no goal clause is
-provided, the default is to perform the conversion only in the
-conclusion.
-
-For backward compatibility, the notation :n:`in {+ @ident}` performs
-the conversion in hypotheses :n:`{+ @ident}`.
-
-.. tacn:: cbv {? @strategy_flag }
-          lazy {? @strategy_flag }
-   :name: cbv; lazy
-
-   These parameterized reduction tactics apply to any goal and perform
-   the normalization of the goal according to the specified flags. In
-   correspondence with the kinds of reduction considered in Coq namely
-   :math:`\beta` (reduction of functional application), :math:`\delta`
-   (unfolding of transparent constants, see :ref:`vernac-controlling-the-reduction-strategies`),
-   :math:`\iota` (reduction of
-   pattern matching over a constructed term, and unfolding of :g:`fix` and
-   :g:`cofix` expressions) and :math:`\zeta` (contraction of local definitions), the
-   flags are either ``beta``, ``delta``, ``match``, ``fix``, ``cofix``,
-   ``iota`` or ``zeta``. The ``iota`` flag is a shorthand for ``match``, ``fix``
-   and ``cofix``. The ``delta`` flag itself can be refined into
-   :n:`delta [ {+ @qualid} ]` or :n:`delta - [ {+ @qualid} ]`, restricting in the first
-   case the constants to unfold to the constants listed, and restricting in the
-   second case the constant to unfold to all but the ones explicitly mentioned.
-   Notice that the ``delta`` flag does not apply to variables bound by a let-in
-   construction inside the :n:`@term` itself (use here the ``zeta`` flag). In
-   any cases, opaque constants are not unfolded (see :ref:`vernac-controlling-the-reduction-strategies`).
-
-   Normalization according to the flags is done by first evaluating the
+   I didn't see how comparing these to "change" was very helpful, so I
+   dropped that sentence, too.
+
+.. tacn:: lazy {? @strategy_flag } {? @occurrences }
+          cbv {? @strategy_flag } {? @occurrences }
+
+   .. insertprodn red_expr pattern_occs
+
+   .. prodn::
+      red_expr ::= lazy {? @strategy_flag }
```

Comment:
I wasn't suggesting moving `red_expr` here. `red_expr` was at its place before any tactic description. What I was suggesting to move was just the `strategy_flag` / `red_flag` / `delta_flag` triplet.

Furthermore, another suggestion of mine was to reorder the productions in `red_expr` to match the order of the tactics described below.

Reply:
I just realized a little while ago that red_expr is not needed here so I moved it into `Declare Reduction`, the only reference to it in the chapter.

Where do you want the (since renamed) strategy_flag et al to appear?

> another suggestion of mine was to reorder the productions in `red_expr` to match the order of the tactics described below.

Which I did in Review #1 and just double-checked.  Did I miss something?

Reply:
> so I moved it into `Declare Reduction`, the only reference to it in the chapter.

Is `Declare Reduction` next to the `eval` tactic?  I think both `red_expr` and `Declare Reduction` really only make sense in the context of `eval` / `Eval`

Reply:
They are currently in other chapters.  Any reference to `red_expr` will be hyperlinked to its grammar and the associated explanation, so the information is easy to discover.  @Zimmi48 what do you think about moving `eval`, `Eval` and `Compute` into this chapter? `Definition` and `rewrite_strat` also refer to `red_expr`, but I think I would leave them where they are.

Reply:
> what do you think about moving `eval`, `Eval` and `Compute` into this chapter?

I'm relatively favorable to this. The only thing that made me hesitant is that currently, all the "query commands" are gathered into a common section. But it could make sense to stop this gathering and to move the general considerations on query commands, e.g. next to the definition of a command.